src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4094,6 +4063,182 @@ void C2_MacroAssembler::vector_castF2I_evex(XMMRegister dst, XMMRegister src, XM
   bind(done);
 }
 
+void C2_MacroAssembler::vector_cast_double_special_cases_evex(XMMRegister dst, XMMRegister src, XMMRegister xtmp1,
```

Comment:
What does this do? Comment, even pseudo code, would be nice.

Reply:
> Hi, IIRC for evex encoding you can embed the RC control bit directly in the evex prefix, removing the need to rely on global MXCSR register. Thanks.

Hi @merykitty ,  You are correct, we can embed RC mode in instruction encoding round instructions (towards -inf,+inf, zero). But to match the semantics of Math.round API one needs to add 0.5[f] to input value and then perform rounding over resultant value, which is why @sviswa7 suggested to use a global rounding mode driven by MXCSR.RC so that intermediate floating inexact values also are resolved as desired, but OOO execution may misplace LDMXCSR and hence may have undesired side effects.

Reply:
> What does this do? Comment, even pseudo code, would be nice.

Thanks @theRealAph , I shall append the comments over the routine.
BTW, entire rounding algorithm can also be implemented using  Vector API which can perform if-conversion using masked operations.
```
class roundf {
   public static VectorSpecies ISPECIES = IntVector.SPECIES_512;
   public static VectorSpecies SPECIES = FloatVector.SPECIES_512;

   public static int round_vector(float[] a, int[] r, int ctr) {
      IntVector shiftVBC = (IntVector) ISPECIES.broadcast(24 - 2 + 127);
      for (int i = 0; i < a.length; i += SPECIES.length()) {
         FloatVector fv = FloatVector.fromArray(SPECIES, a, i);
         IntVector iv = fv.reinterpretAsInts();
         IntVector biasedExpV = iv.lanewise(VectorOperators.AND, 0x7F800000);
         biasedExpV = biasedExpV.lanewise(VectorOperators.ASHR, 23);
         IntVector shiftV = shiftVBC.lanewise(VectorOperators.SUB, biasedExpV);
         VectorMask cond = shiftV.lanewise(VectorOperators.AND, -32)
               .compare(VectorOperators.EQ, 0);
         IntVector res = iv.lanewise(VectorOperators.AND, 0x007FFFFF)
               .lanewise(VectorOperators.OR, 0x007FFFFF + 1);
         VectorMask cond1 = iv.compare(VectorOperators.LT, 0);
         VectorMask cond2 = cond1.and(cond);
         res = res.lanewise(VectorOperators.NEG, cond2);
         res = res.lanewise(VectorOperators.ASHR, shiftV)
               .lanewise(VectorOperators.ADD, 1)
               .lanewise(VectorOperators.ASHR, 1);
         res = fv.convert(VectorOperators.F2I, 0)
               .reinterpretAsInts()
               .blend(res, cond);
         res.intoArray(r, i);
      }
      return r[ctr];
   }
 
```

Reply:
That pseudocode would make a very useful comment too. This whole patch is very thinly commented.

Reply:
> > Hi, IIRC for evex encoding you can embed the RC control bit directly in the evex prefix, removing the need to rely on global MXCSR register. Thanks.
> 
> Hi @merykitty , You are correct, we can embed RC mode in instruction encoding of round instruction (towards -inf,+inf, zero). But to match the semantics of Math.round API one needs to add 0.5[f] to input value and then perform rounding over resultant value, which is why @sviswa7 suggested to use a global rounding mode driven by MXCSR.RC so that intermediate floating inexact values are resolved as desired, but OOO execution may misplace LDMXCSR and hence may have undesired side effects.

**Just want to correct above statement, LDMXCSR will not re-ordered by OOO backend.**

Reply:
> That pseudocode would make a very useful comment too. This whole patch is very thinly commented.

I have replaced earlier bulky sequence, new sequence is having similar performance but reduction in code may improve inlining behavior.  Added descriptive comments around the special cases.

Reply:
> There are already `RoundFloat`, `RoundDouble`, and `RoundDoubleMode` nodes defined.
> 
> Though `RoundFloat` and `RoundDouble` are legacy nodes used only on x86-32, `RoundDoubleMode` supports multiple rounding modes and is amenable to auto-vectorization.
> 
> What do you think about the following alternative?
> 
> Reuse `RoundDoubleMode` (with a new rounding mode) and introduce `RoundFloatMode`.
> 
> Special rounding rules is not the only peculiarity of `Math.round()`. It also converts the result to an integral type. It can be represented as `ConvF2I (RoundFloatMode f #rmode)` / `ConvD2L (RoundDoubleMode d #rmode)`. In scalar case, it can be matched as a single AD instruction.
> 
> Auto-vectorizer can then convert it to `VectorCastF2X (RoundFloatModeV vf #rmode)` / `VectorCastD2X (RoundDoubleModeV vd #rmode)` and match it in a similar manner.

Adding new rounding mode to RoundDoubleMode may disturb other targets. match_rule_supported routine operates over Opcodes and currently any target supporting RoundDoubleMode generates code for all the rounding modes. Your solution is anyways based on creating new scalar and vector IR node for floating point rounding operation, which is what patch is doing currently.